source/magpie/helper.d

```diff
@@ -90,6 +90,34 @@ void sortIndex(string[] index, int[] codes)
     }
 }
 
+/// Does exactly what generateCode does in Index but for any general ndarray
+int[] vectorize(T)(T[] values)
+{
+    int[T] elementPos;
+    int[] pos;
+    int totalUnique = 0;
+
+    import core.exception: RangeError;
+    foreach(i; values)
+    {
+        try
```

Comment:
Please see 
https://dlang.org/spec/hash-map.html#inserting_if_not_present
https://dlang.org/spec/hash-map.html#advanced_updating
and in the [table of properties](https://dlang.org/spec/hash-map.html#properties)
`get` `require` and `update`. and use those instead of using errors(!) for control flow.

Reply:
Thanks for linking the docs. Helped a ton.
I know there was a better way but couldn't find it at that time.
Addressed in dec86d1